### PR TITLE
Update pp save rules to handle time means

### PR DESCRIPTION
**Pull request by hdyson**
As discussed last week.  Most of these additions are novel tests - some to ensure preservation of existing behaviour for cubes without a time mean cell method, and some to exercise the expected behaviour with a time mean cell method.  The actual pp save rules changes are reasonably simple.

Our use case doesn't require changes to pp load rules - so I've not investigated in detail whether a similar change would be needed.  From a cursory look, though, it does look as though pp_load_rules rely on `lbproc` for detecting the cell method: https://github.com/SciTools/iris/blob/master/lib/iris/fileformats/pp_load_rules.py#L964 which in turn is set correctly by the existing save rules.  However, it looks as though there isn't a capability to construct the time coordinate from `t1` and `t2` without also constructing the `forecast_period` and `forecast_reference_time`.  If needed, I can investigate making the reciprocal changes here?

There's also a pile (70+) of legacy flake8 errors in `test_cube_to_pp.py` - I've left these alone to avoid adding noise to the code changes, but I can fix those in passing if needed.

### CODE REVIEW

**Comment by hdyson** [comment]
I think the test failures here already exist on iris/master - I'll rebase once master is fixed?

**Comment by pelson** [comment]
Thanks @hdyson. Looks like we have another case of bit-rot on master, possibly due to a dask change (possibly the same as https://github.com/SciTools/iris/issues/3125, thought I'm not sure).

Thank you for putting this PR together. I've found it really hard to validate the change against the Met Office's F3 document (the rules make it hard to see the wood from the trees, nothing to do with your implementation), and would definitely appreciate some face-to-face discussion (that was super useful last week) about this to help build my confidence that it is doing the right thing. Happy for somebody like @pp-mo / @dkillick to weigh in if they already have the context. I'm out for a couple of weeks, but if this is still awaiting review upon my return, happy to pick this up.

Thanks! 👍

**Comment by hdyson** [comment]
> Thank you for putting this PR together. I've found it really hard to validate the change against the Met Office's F3 document (the rules make it hard to see the wood from the trees, nothing to do with your implementation), and would definitely appreciate some face-to-face discussion

I've attempted to capture the behaviour difference between instantaneous and time mean fields in the tests, so that's probably where I'd start looking rather than at the save rules.  Once you understand what the desired outcomes are, the rule changes may make more sense.  But yes, this is a fairly opaque area, so I'm happy to talk about it.

**Comment by pp-mo** [comment]
Only tangentially connected, but I was tempted to investigate rationalising the logic in pp_save_rules
#3143

**Comment by hdyson** [comment]
@pp-mo I'd suggest that even if the rest of this change isn't incorporated, the half of the tests that are dealing with current behaviour may be useful for #3143 ?

**Comment by pp-mo** [comment]
> @pp-mo I'd suggest that even if the rest of this change isn't incorporated, the half of the tests that are dealing with current behaviour may be useful for #3143 ?

Thanks, that does look useful.  
I have now incorporated your tests + equivalent changes into there in [this latest commit](https://github.com/SciTools/iris/pull/3143/commits/6e6410c0062dfb157fec45b777e6c818d295852a)

Unfortunately I began #3143 as a distraction + I really shouldn't spend more time on it just now ...
I'm minded to merge this now (but see below)
but in reality I'd rather develop + merge the #3143 approach, in view of how hard it now is to understand the rules code.

I wonder if you might consider addressing that ?
It really needs someone to write tests to probe all the code branches in the time rules, then to prove that results are the same as existing (including this, when it goes in).

**Comment by pp-mo** [comment]
Now :+1: I think I'm happy with what this does.

***Except***  I'd much rather the new tests did not go into `iris.tests.test_cube_to_pp`, as that is horribly obsolete in style + not cast as a proper unit test with the usual path rules.
I think this code "should" really live in 'iris.tests.fileformats.pp.test_save`.
Could you arrange that ?

**Comment by hdyson** [comment]
> I think this code "should" really live in 'iris.tests.fileformats.pp.test_save`.
> Could you arrange that ?

Yes, I don't see a problem with doing that.  It's likely to be near the end of the week/early next week when I'm able to pick this up again though, apologies.

I assume I'll need to rebase from master when I pick it up to sort out the random test failures?

**Comment by pp-mo** [comment]
> Yes, I don't see a problem with doing that. It's likely to be near the end of the week/early next week when I'm able to pick this up again though, apologies.

No problem.

> I assume I'll need to rebase from master when I pick it up to sort out the random test failures?

Yes, we fixed that yesterday + finally got tests passing again :lollipop: :champagne:

**Comment by hdyson** [comment]
I think our use case here may be evolving so we do need the corresponding pp load rules changes - I'd prefer to make those changes here too, but if you'd prefer a separate PR please let me know.

File: lib/iris/tests/test_cube_to_pp.py [hunk 1]
```diff
@@ -26,6 +26,7 @@
 import tempfile
 
 import cf_units
+import cftime
```
[the remark this answers is not in the document]
Comment (hdyson) [reply lib/iris/tests/test_cube_to_pp.py]: [This comment](https://github.com/SciTools/iris/blob/ceea6ec23803e461665af761c2242b11c4f5a9b6/lib/iris/tests/unit/fileformats/pp/test_save.py#L170) from an existing test is also the reason I couldn't update the style of these tests to match the other tests.

**Comment by hdyson** [comment]
[This comment](https://github.com/SciTools/iris/blob/ceea6ec23803e461665af761c2242b11c4f5a9b6/lib/iris/tests/unit/fileformats/pp/test_save.py#L170) from an existing test is also the reason I couldn't update the style of these tests to match the other tests.

**Comment by pp-mo** [comment]
I still think we should refactor the rules as-per #3143, 
but more tests are needed to check the equivalence,
which is more work than we can commit to right now ...

**Comment by pp-mo** [comment]
Congrats + thanks @hdyson for engaging with this, this code is not easy to work with !!

**Comment by hdyson** [comment]
Thanks @pp-mo and @pelson , I appreciate you both taking the time on this.

**Pull request merged**